scripts/protocolFilters/fonbnk.ts

```diff
@@ -1,6 +1,82 @@
+import { LogField } from '@envio-dev/hypersync-client'
+import { isAddress, pad } from 'viem'
+import {
+  fonbnkNetworkToNetworkId,
+  TRANSFER_TOPIC,
+} from '../calculateRevenue/protocols/fonbnk/constants'
+import {
+  getFonbnkAssets,
+  getPayoutWallets,
+} from '../calculateRevenue/protocols/fonbnk/helpers'
+import { SUPPORTED_FONBNK_NETWORKS } from '../calculateRevenue/protocols/fonbnk/types'
 import { ReferralEvent } from '../types'
+import { getBlock, getHyperSyncClient } from '../utils'
+import { paginateQuery } from '../utils/hypersyncPagination'
 
-// TODO(ENG-272): Implement filter function
 export async function filter(event: ReferralEvent): Promise<boolean> {
-  return !!event
+  if (!isAddress(event.userAddress)) {
+    throw new Error(`Invalid user address: ${event.userAddress}`)
+  }
+
+  let foundValidTransaction = false
+  let transactedBeforeReferral = false
+
+  const fonbnkAssets = await getFonbnkAssets()
+  // For each network, create a new hypersync client and get all of the unique payout wallets
+  for (const supportedNetwork of SUPPORTED_FONBNK_NETWORKS) {
+    const client = getHyperSyncClient(
+      fonbnkNetworkToNetworkId[supportedNetwork],
+    )
+    const networkAssets = fonbnkAssets
+      .filter((asset) => asset.network === supportedNetwork)
+      .map((asset) => asset.asset)
+    const payoutWallets = await Promise.all(
+      networkAssets.map((asset) =>
+        getPayoutWallets({ fonbnkNetwork: supportedNetwork, asset }),
+      ),
+    )
+    const uniquePayoutWallets = new Set(payoutWallets.flat())
+
+    // For each payout wallet, check that the user not transacted before the referral
+    // and for at least one of the payout wallets, the user has transacted
+    for (const payoutWallet of uniquePayoutWallets) {
```

Comment:
small nit: i might be missing some context here but it seems like you are getting all events from block 0 -> get the transaction -> for each transaction, get the block timestamp (async) -> check if there are events after the known timestamp. would it be more efficient to convert the known timestamp to a block number once per network, then make this query from that block? the check could then be simplified to return true if there any events found? converting a timestamp to a block number isn't very trivial but we have this `getNearestBlock` utility function which may help...

Reply:
OOHHHHH i misunderstood, what you're trying to do is find the _first_ transaction the user made with fonbnk, and verify if it is after the referral was made? in that case, do you need both `transactedBeforeReferral` and `foundValidTransaction`? perhaps you could check only the first transaction returned by the query then to simplify the logic?